### DISCUSSION TITLE
Havoc C2 Agent开发记录 | 倾旋的博客

### DISCUSSION THREAD

**Discussion by giscus** [Rvn0xsy/rvn0xsy.github.io#25] [General]
# Havoc C2 Agent开发记录 | 倾旋的博客

主要简单介绍一下Havoc Framework的跨平台Agent实现过程。

https://payloads.online/archivers/2023/10/12/fda899c9-47ea-40e6-b6fc-a00600ccc9cb

**Comment by Rvn0xsy** [comment]
这个websocket服务可以理解为CobaltStrike的Team server API服务

mengnan ***@***.***> 于2023年12月17日周日 16:00写道：

**Comment by fullfill13** [comment]
为什么我添加agent后无法生成payload文件呢，目前状态是在payload部分有Havoc-Agent这个选项，选中后执行Generate后卡在了[*] Agent Config: {'Sleep': '10'}

**Reply by Rvn0xsy** [reply]
有更具体的log吗？一般情况下因为要混淆生成，会比较消耗服务器的资源，生成的时间较长，你可以观察一下python agent的日志有没有一些错误异常信息

**Comment by destroylq** [comment]
你好，我在最新版的Havoc上测试你的代码失败。原因是与ts的websocket连接时失败，后面也就无法注册agent

**Reply by Rvn0xsy** [reply]
ts的websocket连接功能在main分支给关闭了，你可能需要切换到dev分支，使用dev分支的代码编译ts
https://github.com/HavocFramework/Havoc/tree/dev

**Reply by destroylq** [reply]
感谢，可以给一下websocket连接功能关闭的代码具体位置吗，我没找到

**Reply by destroylq** [reply]
我看到main分支commit里面，重启了service api功能，没看到关于websocket的消息 
https://github.com/HavocFramework/Havoc/commit/e4455a5ff37f7ab6f236fdfa09bbe467896c8bc7

**Reply by Rvn0xsy** [reply]
在Dev分支中： https://github.com/HavocFramework/Havoc/blob/3bf236c417b3963034ae3722f41bf521f14a3476/teamserver/cmd/server/teamserver.go#L192C4-L192C4

**Reply by Kano-2525-a** [reply]
按照Talon的 requirements.txt 安装后出现以下错误。
AttributeError: module 'websocket' has no attribute 'WebSocketApp'. 

谷歌看了下是版本问题？所以我卸载了websocket 0.2.1 .重新安装了websocket-client-1.7.0 websocket-client2-1.0  然后发生了以下错误

[*] Connect to Havoc service api
[-] Websocket error: Handshake status 404 Not Found -+-+- {'content-type': 'text/plain', 'date': 'Wed, 20 Mar 2024 11:24:43 GMT', 'content-length': '18'} -+-+- b'404 page not found'

这也是因为用的最新版本的Havoc吗？感谢

**Reply by Kano-2525-a** [reply]
我解决了这个问题，
在 Talon.py 同目录下，
 git clone https://github.com/HavocFramework/havoc-py/tree/main/havoc
touch havoc/init.py
python3 Talon.py

**Reply by zhihuba** [reply]
@Kano-2525-a 大佬咋解决的。。按照你给的不行哇

**Reply by zhihuba** [reply]
解决了。profiles 中添加 Service {
    Endpoint = "service-endpoint"
    Password = "service-password"
}

**Comment by fullfill13** [comment]
使用http监听器，运行生成的文件后，在ts端会报错，然后自动退出。panic: runtime error: index out of range [4] with length 4
